Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -2629,140 +2629,176 @@ RefPtr<PositionStyleValue> Parser::parse_position_value(TokenStream<ComponentVal
-            } else {
-                auto length_percentage = parse_length_percentage(second_token);
-                if (length_percentage.has_value()) {
-                    transaction_optional_parse.commit();
-                    vertical_edge = EdgeStyleValue::create(PositionEdge::Top, *length_percentage);
-                }
+            if (auto maybe_position = parse_position_edge(token); maybe_position.has_value()) {
+                auto position = maybe_position.release_value();
+                bool valid = as_horizontal ? is_horizontal(position, true) : is_vertical(position, true);
+                if (!valid)
+                    return nullptr;
+                return make_edge_style_value(position, as_horizontal);
             }
-        }
 
-        transaction.commit();
+            auto maybe_length = parse_length_percentage(token);
+            if (!maybe_length.has_value())
+                return nullptr;
+
+            return EdgeStyleValue::create(as_horizontal ? PositionEdge::Left : PositionEdge::Top, maybe_length.release_value());
+        };
+
+        // [ left | center | right | <length-percentage> ]
+        RefPtr<EdgeStyleValue> horitontal_edge = parse_position_or_length(true);
+        if (!horitontal_edge)
+            return nullptr;
+
+        // [ top | center | bottom | <length-percentage> ]
+        RefPtr<EdgeStyleValue> vertical_edge = parse_position_or_length(false);
         if (!vertical_edge)
-            vertical_edge = make_edge_style_value(PositionEdge::Center, false);
-        return PositionStyleValue::create(horizontal_edge.release_nonnull(), vertical_edge.release_nonnull());
+            return nullptr;
+
+        transaction.commit();
+        return PositionStyleValue::create(horitontal_edge.release_nonnull(), vertical_edge.release_nonnull());
     };
 
     // [ [ left | right ] <length-percentage> ] &&
     // [ [ top | bottom ] <length-percentage> ]
-    auto alternative_3 = [&]() -> RefPtr<PositionStyleValue> {
-        auto transaction = tokens.begin_transaction();
-        tokens.skip_whitespace();
-        RefPtr<EdgeStyleValue> horizontal_edge;
-        RefPtr<EdgeStyleValue> vertical_edge;
+    auto alternative_4 = [&]() -> RefPtr<PositionStyleValue> {
+        struct PositionAndLength {
+            PositionEdge position;
+            LengthPercentage length;
+        };
 
-        auto parse_horizontal = [&] {
-            // [ left | right ] <length-percentage> ]
-            auto transaction = tokens.begin_transaction();
+        auto parse_position_and_length = [&]() -> Optional<PositionAndLength> {
             tokens.skip_whitespace();
-            auto edge = parse_position_edge(tokens.next_token());
-            if (!edge.has_value() || !is_horizontal(*edge, false))
-                return false;
+            auto first_token = tokens.next_token();
+
+            auto maybe_position = parse_position_edge(first_token);
```

Comment:
Here we can avoid the `first_token` variable altogether and just pass `tokens.next_token()` to `parse_position_edge()`.